api/agent/drivers/docker/cookie.go

```diff
@@ -298,19 +301,31 @@ func (c *cookie) ValidateImage(ctx context.Context) (bool, error) {
 	c.imgAuthConf = config
 
 	// see if we already have it
-	img, err := c.drv.docker.InspectImage(ctx, c.task.Image())
+	img, err := c.inspectImage(ctx, c.task.Image())
 	if err == docker.ErrNoSuchImage {
 		return true, nil
 	}
+
 	// we have the image, notify image cache
-	if err == nil && c.drv.imgCache != nil && c.cachedImg == nil {
-		c.cachedImg = &CachedImage{
-			ID:       img.ID,
-			ParentID: img.Parent,
-			RepoTags: img.RepoTags,
-			Size:     uint64(img.Size),
+	if c.drv.imgCache != nil && c.cachedImg == nil {
+		if err == nil {
+			c.cachedImg = &CachedImage{
+				ID:       img.ID,
+				ParentID: img.Parent,
+				RepoTags: img.RepoTags,
+				Size:     uint64(img.Size),
+			}
+			c.drv.imgCache.MarkBusy(c.cachedImg)
+		}
+		if err == ErrImageWithVolume {
```

Comment:
assuming we delete images from disk that we pull only to find volumes, shouldn't we just never add this to the image cache? the previous error condition seems fine in this respect, but perhaps i'm missing the ordering of pulling and adding to image cache

Reply:
Volume checker here leaves the cleanup to image cacher. If a volume is in the image, then it is inserted into LRU via Update() and if the need to cleanup ever arises then we assume LRU would kick in and clean this up. (Note that LRU Update() is not perfect here as this code path might keep this image at the head of the cache. I might provide a special image cacher function instead of Update() in the near future as an improvement.)